File: sdks/typescript/src/testing.ts
/**
 * Contd SDK Testing Utilities
 * Test harnesses and mocks for workflow testing
 */

import { WorkflowState, BaseEvent } from './types';
import { WorkflowInterrupted } from './errors';
import { ExecutionContext, ExecutionEngine, Lease } from './context';
import { WorkflowRegistry } from './registry';

// ============================================================================
// Step and Workflow Execution Records
// ============================================================================

export interface StepExecution {
  stepName: string;
  stepId: string;
  attempt: number;
  startedAt: Date;
  completedAt?: Date;
  durationMs: number;
  result?: unknown;
  error?: string;
  wasCached: boolean;
}

export interface WorkflowExecution {
  workflowId: string;
  workflowName: string;
  startedAt: Date;
  completedAt?: Date;
  status: 'running' | 'completed' | 'interrupted' | 'failed';
  steps: StepExecution[];
  finalState?: WorkflowState;
  error?: string;
  interruptedAtStep?: number;
}

// ============================================================================
// Mock Execution Engine
// ============================================================================

export class MockExecutionEngine implements ExecutionEngine {
  private _interruptAtStep?: number;
  private _failAtStep?: number;
  private _failWith?: new (message: string) => Error;
  private _recordedEvents: BaseEvent[] = [];
  private _stepCounter = 0;
  private _states: Map<string, WorkflowState> = new Map();
  private _completedSteps: Map<string, WorkflowState> = new Map();

  leaseManager = {
    acquire: async (workflowId: string, ownerId: string): Promise<Lease | null> => {
      return {
        workflowId,
        ownerId,
        expiresAt: new Date(Date.now() + 60000),
      };
    },
    release: async (_lease: Lease): Promise<void> => {},
    heartbeat: async (_lease: Lease): Promise<void> => {},
    HEARTBEAT_INTERVAL: 10000,
  };

  journal = {
    append: async (event: unknown): Promise<void> => {
      this._recordedEvents.push(event as BaseEvent);
    },
  };

  idempotency = {
    checkCompleted: async (
      workflowId: string,
      stepId: string
    ): Promise<WorkflowState | null> => {
      return this._completedSteps.get(`${workflowId}:${stepId}`) || null;
    },
    allocateAttempt: async (
      _workflowId: string,
      _stepId: string,
      _lease: Lease | null
    ): Promise<number> => {
      return ++this._stepCounter;
    },
    markCompleted: async (
      workflowId: string,
      stepId: string,
      _attemptId: number,
      state: WorkflowState
    ): Promise<void> => {
      this._completedSteps.set(`${workflowId}:${stepId}`, state);
    },
  };

  async restore(workflowId: string): Promise<WorkflowState> {
    const state = this._states.get(workflowId);
    if (!state) {
      return {
        workflowId,
        stepNumber: 0,
        variables: {},
        metadata: {},
        version: '1.0',
        checksum: '',
        orgId: 'default',
      };
    }
    return state;
  }

  async completeWorkflow(workflowId: string): Promise<void> {
    const state = this._states.get(workflowId);
    if (state) {
      state.metadata.completedAt = new Date().toISOString();
    }
  }

  async maybeSnapshot(state: WorkflowState): Promise<void> {
    this._states.set(state.workflowId, state);
  }

  /**
   * Configure interruption at specific step
   */
  setInterruptAt(stepNumber: number): void {
    this._interruptAtStep = stepNumber;
  }

  /**
   * Configure failure injection at specific step
   */
  setFailAt(
    stepNumber: number,
    exceptionType: new (message: string) => Error = Error
  ): void {
    this._failAtStep = stepNumber;
    this._failWith = exceptionType;
  }

  /**
   * Check if workflow should be interrupted
   */
  checkInterrupt(stepNumber: number, workflowId: string): void {
    if (
      this._interruptAtStep !== undefined &&
      stepNumber >= this._interruptAtStep
    ) {
      throw new WorkflowInterrupted(workflowId, stepNumber);
    }
  }

  /**
   * Check if failure should be injected
   */
  checkFailure(stepNumber: number): void {
    if (this._failAtStep !== undefined && stepNumber === this._failAtStep) {
      const ErrorClass = this._failWith || Error;
      throw new ErrorClass(`Injected failure at step ${stepNumber}`);
    }
  }

  /**
   * Get recorded events
   */
  getRecordedEvents(): BaseEvent[] {
    return [...this._recordedEvents];
  }

  /**
   * Clear recorded events
   */
  clearRecordedEvents(): void {
    this._recordedEvents = [];
  }

  /**
   * Reset all mock state
   */
  reset(): void {
    this._interruptAtStep = undefined;
    this._failAtStep = undefined;
    this._failWith = undefined;
    this._recordedEvents = [];
    this._stepCounter = 0;
    this._states.clear();
    this._completedSteps.clear();
  }
}

// ============================================================================
// Test Case Helper
// ============================================================================

export class ContdTestCase {
  public engine: MockExecutionEngine;
  public executions: WorkflowExecution[] = [];
  public currentExecution?: WorkflowExecution;

  constructor() {
    this.engine = new MockExecutionEngine();
  }

  /**
   * Set up test fixtures
   */
  setUp(): void {
    this.engine.reset();
    this.executions = [];
    this.currentExecution = undefined;
  }

  /**
   * Tear down test fixtures
   */
  tearDown(): void {
    this.engine.reset();
  }

  /**
   * Run a workflow with optional interruption or failure injection
   */
  async runWorkflow<T>(
    workflowFn: (...args: unknown[]) => Promise<T>,
    options: {
      args?: unknown[];
      interruptAtStep?: number;
      failAtStep?: number;
      failWith?: new (message: string) => Error;
    } = {}
  ): Promise<T | undefined> {
    // Configure mock engine
    if (options.interruptAtStep !== undefined) {
      this.engine.setInterruptAt(options.interruptAtStep);
    }
    if (options.failAtStep !== undefined) {
      this.engine.setFailAt(options.failAtStep, options.failWith);
    }

    // Create execution record
    const execution: WorkflowExecution = {
      workflowId: '',
      workflowName: workflowFn.name,
      startedAt: new Date(),
      status: 'running',
      steps: [],
    };
    this.currentExecution = execution;
    this.executions.push(execution);

    try {
      const result = await workflowFn(...(options.args || []));
      execution.status = 'completed';
      execution.completedAt = new Date();
      return result;
    } catch (error) {
      if (error instanceof WorkflowInterrupted) {
        execution.status = 'interrupted';
        execution.interruptedAtStep = error.details.interruptedAtStep as number;
        return undefined;
      }
      execution.status = 'failed';
      execution.error = (error as Error).message;
      execution.completedAt = new Date();
      throw error;
    }
  }

  /**
   * Resume an interrupted workflow
   */
  async resumeWorkflow<T>(
    workflowFn?: (...args: unknown[]) => Promise<T>,
    options: { args?: unknown[] } = {}
  ): Promise<T | undefined> {
    // Clear interrupt setting
    this.engine.reset();

    if (!workflowFn && this.currentExecution) {
      workflowFn = WorkflowRegistry.get(
        this.currentExecution.workflowName
      ) as typeof workflowFn;
    }

    if (!workflowFn) {
      throw new Error('No workflow function provided');
    }

    return this.runWorkflow(workflowFn, { args: options.args });
  }

  // ========================================================================
  // Assertions
  // ========================================================================

  assertCompleted(message = ''): void {
    if (!this.currentExecution) {
      throw new Error('No workflow execution to check');
    }
    if (this.currentExecution.status !== 'completed') {
      throw new Error(
        `Workflow not completed: status=${this.currentExecution.status}. ${message}`
      );
    }
  }

  assertInterrupted(atStep?: number, message = ''): void {
    if (!this.currentExecution) {
      throw new Error('No workflow execution to check');
    }
    if (this.currentExecution.status !== 'interrupted') {
      throw new Error(
        `Workflow not interrupted: status=${this.currentExecution.status}. ${message}`
      );
    }
    if (
      atStep !== undefined &&
      this.currentExecution.interruptedAtStep !== atStep
    ) {
      throw new Error(
        `Interrupted at wrong step: expected=${atStep}, actual=${this.currentExecution.interruptedAtStep}. ${message}`
      );
    }
  }

  assertFailed(errorContains?: string, message = ''): void {
    if (!this.currentExecution) {
      throw new Error('No workflow execution to check');
    }
    if (this.currentExecution.status !== 'failed') {
      throw new Error(
        `Workflow not failed: status=${this.currentExecution.status}. ${message}`
      );
    }
    if (
      errorContains &&
      !this.currentExecution.error?.includes(errorContains)
    ) {
      throw new Error(
        `Error message doesn't contain '${errorContains}': actual='${this.currentExecution.error}'. ${message}`
      );
    }
  }

  assertStepCount(expected: number, message = ''): void {
    if (!this.currentExecution) {
      throw new Error('No workflow execution to check');
    }
    const actual = this.currentExecution.steps.length;
    if (actual !== expected) {
      throw new Error(
        `Step count mismatch: expected=${expected}, actual=${actual}. ${message}`
      );
    }
  }

  assertEventCount(expected: number, eventType?: string, message = ''): void {
    let events = this.engine.getRecordedEvents();
    if (eventType) {
      events = events.filter((e) => e.eventType === eventType);
    }
    if (events.length !== expected) {
      throw new Error(
        `Event count mismatch: expected=${expected}, actual=${events.length}. ${message}`
      );
    }
  }

  // ========================================================================
  // Utilities
  // ========================================================================

  getEvents(eventType?: string): BaseEvent[] {
    let events = this.engine.getRecordedEvents();
    if (eventType) {
      events = events.filter((e) => e.eventType === eventType);
    }
    return events;
  }

  getFinalState(): WorkflowState | undefined {
    return this.currentExecution?.finalState;
  }

  printExecutionSummary(): void {
    if (!this.currentExecution) {
      console.log('No execution to summarize');
      return;
    }

    const ex = this.currentExecution;
    console.log('\n' + '='.repeat(50));
    console.log(`Workflow: ${ex.workflowName}`);
    console.log(`Status: ${ex.status}`);
    console.log(`Steps: ${ex.steps.length}`);
    if (ex.interruptedAtStep !== undefined) {
      console.log(`Interrupted at: step ${ex.interruptedAtStep}`);
    }
    if (ex.error) {
      console.log(`Error: ${ex.error}`);
    }
    console.log(`Events recorded: ${this.engine.getRecordedEvents().length}`);
    console.log('='.repeat(50) + '\n');
  }
}

// ============================================================================
// Fluent Test Builder
// ============================================================================

export class WorkflowTestBuilder<T> {
  private workflowFn: (...args: unknown[]) => Promise<T>;
  private testCase: ContdTestCase;
  private _args: unknown[] = [];
  private _interruptAt?: number;
  private _failAt?: number;
  private _failWith?: new (message: string) => Error;

  constructor(workflowFn: (...args: unknown[]) => Promise<T>) {
    this.workflowFn = workflowFn;
    this.testCase = new ContdTestCase();
    this.testCase.setUp();
  }

  withInput(...args: unknown[]): this {
    this._args = args;
    return this;
  }

  interruptAt(step: number): this {
    this._interruptAt = step;
    return this;
  }

  failAt(step: number, exception?: new (message: string) => Error): this {
    this._failAt = step;
    this._failWith = exception;
    return this;
  }

  async run(): Promise<this> {
    await this.testCase.runWorkflow(this.workflowFn, {
      args: this._args,
      interruptAtStep: this._interruptAt,
      failAtStep: this._failAt,
      failWith: this._failWith,
    });
    this._interruptAt = undefined;
    this._failAt = undefined;
    return this;
  }

  async resume(): Promise<this> {
    await this.testCase.resumeWorkflow(this.workflowFn, { args: this._args });
    return this;
  }

  assertCompleted(): this {
    this.testCase.assertCompleted();
    return this;
  }

  assertInterrupted(atStep?: number): this {
    this.testCase.assertInterrupted(atStep);
    return this;
  }

  assertFailed(errorContains?: string): this {
    this.testCase.assertFailed(errorContains);
    return this;
  }

  cleanup(): void {
    this.testCase.tearDown();
  }
}

File: sdks/go/testing.go
package contd

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
)

// StepExecution records a step execution during testing
type StepExecution struct {
	StepName    string
	StepID      string
	Attempt     int
	StartedAt   time.Time
	CompletedAt *time.Time
	DurationMs  int64
	Result      interface{}
	Error       string
	WasCached   bool
}

// WorkflowExecution records a workflow execution during testing
type WorkflowExecution struct {
	WorkflowID        string
	WorkflowName      string
	StartedAt         time.Time
	CompletedAt       *time.Time
	Status            string
	Steps             []StepExecution
	FinalState        *WorkflowState
	Error             string
	InterruptedAtStep *int
}

// MockEngine is a mock execution engine for testing
type MockEngine struct {
	mu              sync.RWMutex
	interruptAtStep *int
	failAtStep      *int
	failWith        error
	recordedEvents  []interface{}
	stepCounter     int
	states          map[string]*WorkflowState
	completedSteps  map[string]*WorkflowState

	leaseManager      *MockLeaseManager
	journal           *MockJournal
	idempotencyMgr    *MockIdempotencyManager
}

// NewMockEngine creates a new mock engine
func NewMockEngine() *MockEngine {
	engine := &MockEngine{
		recordedEvents: make([]interface{}, 0),
		states:         make(map[string]*WorkflowState),
		completedSteps: make(map[string]*WorkflowState),
	}
	engine.leaseManager = &MockLeaseManager{engine: engine}
	engine.journal = &MockJournal{engine: engine}
	engine.idempotencyMgr = &MockIdempotencyManager{engine: engine}
	return engine
}

// Restore restores workflow state
func (e *MockEngine) Restore(workflowID string) (*WorkflowState, error) {
	e.mu.RLock()
	defer e.mu.RUnlock()
	if state, ok := e.states[workflowID]; ok {
		return state, nil
	}
	return &WorkflowState{
		WorkflowID: workflowID,
		StepNumber: 0,
		Variables:  make(map[string]interface{}),
		Metadata:   make(map[string]interface{}),
		Version:    "1.0",
		OrgID:      "default",
	}, nil
}

// CompleteWorkflow marks a workflow as complete
func (e *MockEngine) CompleteWorkflow(workflowID string) error {
	e.mu.Lock()
	defer e.mu.Unlock()
	if state, ok := e.states[workflowID]; ok {
		state.Metadata["completed_at"] = time.Now().UTC().Format(time.RFC3339)
	}
	return nil
}

// MaybeSnapshot stores a snapshot
func (e *MockEngine) MaybeSnapshot(state *WorkflowState) error {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.states[state.WorkflowID] = state
	return nil
}

// LeaseManager returns the lease manager
func (e *MockEngine) LeaseManager() LeaseManager {
	return e.leaseManager
}

// Journal returns the journal
func (e *MockEngine) Journal() Journal {
	return e.journal
}

// Idempotency returns the idempotency manager
func (e *MockEngine) Idempotency() IdempotencyManager {
	return e.idempotencyMgr
}

// SetInterruptAt configures interruption at a specific step
func (e *MockEngine) SetInterruptAt(stepNumber int) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.interruptAtStep = &stepNumber
}

// SetFailAt configures failure injection at a specific step
func (e *MockEngine) SetFailAt(stepNumber int, err error) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.failAtStep = &stepNumber
	e.failWith = err
}

// CheckInterrupt checks if workflow should be interrupted
func (e *MockEngine) CheckInterrupt(stepNumber int, workflowID string) error {
	e.mu.RLock()
	defer e.mu.RUnlock()
	if e.interruptAtStep != nil && stepNumber >= *e.interruptAtStep {
		return NewWorkflowInterrupted(workflowID, stepNumber)
	}
	return nil
}

// CheckFailure checks if failure should be injected
func (e *MockEngine) CheckFailure(stepNumber int) error {
	e.mu.RLock()
	defer e.mu.RUnlock()
	if e.failAtStep != nil && stepNumber == *e.failAtStep {
		if e.failWith != nil {
			return e.failWith
		}
		return fmt.Errorf("injected failure at step %d", stepNumber)
	}
	return nil
}

// GetRecordedEvents returns recorded events
func (e *MockEngine) GetRecordedEvents() []interface{} {
	e.mu.RLock()
	defer e.mu.RUnlock()
	result := make([]interface{}, len(e.recordedEvents))
	copy(result, e.recordedEvents)
	return result
}

// ClearRecordedEvents clears recorded events
func (e *MockEngine) ClearRecordedEvents() {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.recordedEvents = make([]interface{}, 0)
}

// Reset resets all mock state
func (e *MockEngine) Reset() {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.interruptAtStep = nil
	e.failAtStep = nil
	e.failWith = nil
	e.recordedEvents = make([]interface{}, 0)
	e.stepCounter = 0
	e.states = make(map[string]*WorkflowState)
	e.completedSteps = make(map[string]*WorkflowState)
}

// MockLeaseManager is a mock lease manager
type MockLeaseManager struct {
	engine *MockEngine
}

func (m *MockLeaseManager) Acquire(workflowID, ownerID string) (*Lease, error) {
	return &Lease{
		WorkflowID: workflowID,
		OwnerID:    ownerID,
		ExpiresAt:  time.Now().Add(time.Minute),
	}, nil
}

func (m *MockLeaseManager) Release(lease *Lease) error {
	return nil
}

func (m *MockLeaseManager) Heartbeat(lease *Lease) error {
	return nil
}

func (m *MockLeaseManager) HeartbeatInterval() time.Duration {
	return 10 * time.Second
}

// MockJournal is a mock journal
type MockJournal struct {
	engine *MockEngine
}

func (m *MockJournal) Append(event interface{}) error {
	m.engine.mu.Lock()
	defer m.engine.mu.Unlock()
	m.engine.recordedEvents = append(m.engine.recordedEvents, event)
	return nil
}

// MockIdempotencyManager is a mock idempotency manager
type MockIdempotencyManager struct {
	engine *MockEngine
}

func (m *MockIdempotencyManager) CheckCompleted(workflowID, stepID string) (*WorkflowState, error) {
	m.engine.mu.RLock()
	defer m.engine.mu.RUnlock()
	key := fmt.Sprintf("%s:%s", workflowID, stepID)
	return m.engine.completedSteps[key], nil
}

func (m *MockIdempotencyManager) AllocateAttempt(workflowID, stepID string, lease *Lease) (int, error) {
	m.engine.mu.Lock()
	defer m.engine.mu.Unlock()
	m.engine.stepCounter++
	return m.engine.stepCounter, nil
}

func (m *MockIdempotencyManager) MarkCompleted(workflowID, stepID string, attemptID int, state *WorkflowState) error {
	m.engine.mu.Lock()
	defer m.engine.mu.Unlock()
	key := fmt.Sprintf("%s:%s", workflowID, stepID)
	m.engine.completedSteps[key] = state
	return nil
}

// TestCase is a test harness for workflow testing
type TestCase struct {
	Engine           *MockEngine
	Executions       []WorkflowExecution
	CurrentExecution *WorkflowExecution
}

// NewTestCase creates a new test case
func NewTestCase() *TestCase {
	return &TestCase{
		Engine:     NewMockEngine(),
		Executions: make([]WorkflowExecution, 0),
	}
}

// SetUp sets up test fixtures
func (tc *TestCase) SetUp() {
	tc.Engine.Reset()
	tc.Executions = make([]WorkflowExecution, 0)
	tc.CurrentExecution = nil
}

// TearDown tears down test fixtures
func (tc *TestCase) TearDown() {
	tc.Engine.Reset()
}

// RunWorkflowOptions contains options for running a workflow in tests
type RunWorkflowOptions struct {
	Input           interface{}
	InterruptAtStep *int
	FailAtStep      *int
	FailWith        error
}

// RunWorkflow runs a workflow with optional interruption or failure injection
func (tc *TestCase) RunWorkflow(ctx context.Context, workflowName string, fn WorkflowFunc, opts RunWorkflowOptions) (interface{}, error) {
	// Configure mock engine
	if opts.InterruptAtStep != nil {
		tc.Engine.SetInterruptAt(*opts.InterruptAtStep)
	}
	if opts.FailAtStep != nil {
		tc.Engine.SetFailAt(*opts.FailAtStep, opts.FailWith)
	}

	// Create execution record
	execution := WorkflowExecution{
		WorkflowID:   "wf-" + uuid.New().String(),
		WorkflowName: workflowName,
		StartedAt:    time.Now(),
		Status:       "running",
		Steps:        make([]StepExecution, 0),
	}
	tc.CurrentExecution = &execution
	tc.Executions = append(tc.Executions, execution)

	// Run workflow
	runner := NewWorkflowRunner(tc.Engine, WorkflowConfig{})
	result, err := runner.Run(ctx, workflowName, fn, opts.Input)

	if err != nil {
		if _, ok := err.(*WorkflowInterrupted); ok {
			execution.Status = "interrupted"
			if ie, ok := err.(*WorkflowInterrupted); ok {
				execution.InterruptedAtStep = &ie.StepNumber
			}
			return nil, nil
		}
		execution.Status = "failed"
		execution.Error = err.Error()
		now := time.Now()
		execution.CompletedAt = &now
		return nil, err
	}

	execution.Status = "completed"
	now := time.Now()
	execution.CompletedAt = &now
	return result, nil
}

// ResumeWorkflow resumes an interrupted workflow
func (tc *TestCase) ResumeWorkflow(ctx context.Context, workflowName string, fn WorkflowFunc, input interface{}) (interface{}, error) {
	tc.Engine.Reset()
	return tc.RunWorkflow(ctx, workflowName, fn, RunWorkflowOptions{Input: input})
}

// AssertCompleted asserts that the last workflow completed
func (tc *TestCase) AssertCompleted() error {
	if tc.CurrentExecution == nil {
		return fmt.Errorf("no workflow execution to check")
	}
	if tc.CurrentExecution.Status != "completed" {
		return fmt.Errorf("workflow not completed: status=%s", tc.CurrentExecution.Status)
	}
	return nil
}

// AssertInterrupted asserts that the last workflow was interrupted
func (tc *TestCase) AssertInterrupted(atStep *int) error {
	if tc.CurrentExecution == nil {
		return fmt.Errorf("no workflow execution to check")
	}
	if tc.CurrentExecution.Status != "interrupted" {
		return fmt.Errorf("workflow not interrupted: status=%s", tc.CurrentExecution.Status)
	}
	if atStep != nil && (tc.CurrentExecution.InterruptedAtStep == nil || *tc.CurrentExecution.InterruptedAtStep != *atStep) {
		return fmt.Errorf("interrupted at wrong step: expected=%d, actual=%v", *atStep, tc.CurrentExecution.InterruptedAtStep)
	}
	return nil
}

// AssertFailed asserts that the last workflow failed
func (tc *TestCase) AssertFailed(errorContains string) error {
	if tc.CurrentExecution == nil {
		return fmt.Errorf("no workflow execution to check")
	}
	if tc.CurrentExecution.Status != "failed" {
		return fmt.Errorf("workflow not failed: status=%s", tc.CurrentExecution.Status)
	}
	if errorContains != "" && tc.CurrentExecution.Error == "" {
		return fmt.Errorf("error message doesn't contain '%s': actual='%s'", errorContains, tc.CurrentExecution.Error)
	}
	return nil
}

// GetEvents returns recorded events
func (tc *TestCase) GetEvents(eventType string) []interface{} {
	events := tc.Engine.GetRecordedEvents()
	if eventType == "" {
		return events
	}
	filtered := make([]interface{}, 0)
	for _, e := range events {
		if m, ok := e.(map[string]interface{}); ok {
			if m["event_type"] == eventType {
				filtered = append(filtered, e)
			}
		}
	}
	return filtered
}
